container/prune_tar.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2015 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This tool build tar files from a container_layer,
+while excluding some files and directories."""
+
+import gflags
+import json
+import sys
+
+from container import archive
+
+gflags.DEFINE_string('output', None, 'The output file, mandatory')
+gflags.MarkFlagAsRequired('output')
+
+gflags.DEFINE_string(
+    'manifest', None, 'JSON manifest of contents to remove from the layer')
+
+gflags.DEFINE_string(
+    'root_directory', '.', 'Default root directory is named "."'
+    'Windows docker images require this be named "Files" instead of "."')
+
+FLAGS = gflags.FLAGS
+
+
+class TarFile(object):
+  """A class to generate a pruned container layer."""
+
+  def __init__(self, output, root_directory, tar, remove_paths):
+    self.output = output
+    self.root_directory = root_directory
+    self.tar = tar
+    self.remove_paths = remove_paths
+
+  def __enter__(self):
+    self.tarfile = archive.TarFileWriter(self.output)
+    return self
+
+  def __exit__(self, t, v, traceback):
+    self.tarfile.close()
+
+  def name_filter(self, path):
+    """Callback for archive.TarFileWriter.add_tar that returns false
+    if the file or directory in remove_paths.
+
+    Args:
+       path: the path to the directory or file in the tar
+    """
+    if path.startswith(self.root_directory + '/'):
+      path = path[len(self.root_directory):]
+
+    for remove_path in self.remove_paths:
+      if remove_path.endswith('/'):
+        if path.startswith(remove_path):
+          return False
+      elif path == remove_path or path.startswith(remove_path + '/'):
+        return False
+
+    return True
+
+  def process_tar(self):
+    """Merge a tar file into the destination tar file,
+    while excluding files and directories in remove_paths.
+    All paths in remove_paths must start with a slash (/).
+    When the remove path ends with a slash (/), the directory is preserved
+    while the contents inside the directory is removed.
+    When the remove path does not end with a slash, the file or directory
+    including its contents, is removed.
+    """
+    self.tarfile.add_tar(self.tar, name_filter=self.name_filter)
+
+
+def main(unused_argv):
+  if FLAGS.manifest:
+    with open(FLAGS.manifest, 'r') as f:
+      manifest = json.load(f)
+      remove_paths = manifest.get('remove_paths', [])
```

Comment:
I'm assuming many users might also want to use glob here. It might be better to support that instead of the slightly non-obvious semantic currently supported, i.e. if the path ends in "/" it deletes the directory contents but preserves the directory itself.

Supporting glob should be trivial with https://docs.python.org/3/library/fnmatch.html#module-fnmatch I think?

Reply:
Great idea. Let me find a way to support glob. fnmatch doesn't seem to be the right way because

```
Note that the filename separator ('/' on Unix) is not special to this module. See module glob for pathname expansion (glob uses filter() to match pathname segments). Similarly, filenames starting with a period are not special for this module, and are matched by the * and ? patterns.
```